tail: fix GNU test 'misc/tail'

bbara opened the pull request:
Hi!

This PR targets compatibility to `gnu/tests/misc/tail.pl`.
It's not fully achieved (yet) as some error messages don't fit 1:1 (e.g. a couple of additional `Try 'tail --help' for more information.`). Most of the code is to work around clap, which interprets a couple of arguments as input files.

I imported most of the perl tests to the code base.

Please feel free to respond if I should handle something differently (add tests/comments, split commit, ...) or if this is not needed at all :smile: 

Best regards
bb

sylvestre commented on src/uu/tail/src/args.rs:
```diff
@@ -208,6 +220,62 @@ impl Settings {
+            const DEFAULT_INTVAL: u64 = 10;
+            let mut name = inputs[0].display_name.to_string();
+            let mut default_mode = false;
+            if let Some(mut mode) = name.chars().last() {
```
maybe move most this into a separate function

bbara replied:
moved to FilterMode impl, I guess it fits better there

sylvestre commented on tests/by-util/test_tail.rs:
```diff
@@ -4479,3 +4479,264 @@ fn test_args_sleep_interval_when_illegal_argument_then_usage_error() {
 }
+
+#[test]
+fn test_gnu() {
```
could you please split this test for what it is testing?
easier to debug if we are regressing

bbara replied:
i split them up into argument "type" (-c, +c, -l, +l, -b, -${num}, +${num}) and errors. removed the obvious tests that are already working on main

tertsdiepraam commented on src/uu/tail/src/args.rs:
```diff
@@ -479,6 +540,20 @@ pub fn uu_app() -> Command {
                 .num_args(1..)
                 .value_hint(clap::ValueHint::FilePath),
         )
+        .arg(
```
I'm not sure about doing this with clap. These characters seem to be part of the obsolete syntax right? Couldn't we just fix the parsing for that and get it to work without additional flags? If not, then this is fine, but I think it'd be good to have a bit more comments to explain what these flags are for and how this parsing works.

bbara replied:
Yes, seems to be pretty obsolete.

The problem is, that calling `tail -l <file>` should behave like `tail -10l <file>`, which should basically behave like `tail -n 10 <file>` (`tail -b <file>` is similar). Basically, the flags are introduced only for these two corner cases.

Otherwise, clap is "stealing" them, and I am not sure how to work around it:
```
error: Found argument '-l' which wasn't expected, or isn't valid in this context

  If you tried to supply '-l' as a value rather than a flag, use '-- -l'
```

bbara replied:
So I can add a comment like "/* required for corner case: tail -l <file>*/", if wanted

tertsdiepraam replied:
Ah sorry, I think I was being too vague. With obsolete syntax I mean the `-10` thing, which is parsed by this function: https://github.com/uutils/coreutils/blob/816e0d51e35697104f59427f4a85e77ae5f2c6fd/src/uu/tail/src/parse.rs#L15

The `l` and `c` are part of that argument, instead of being separate arguments, so it seems that we just have to fix this function to support `l` and `c`. We call this function before clap, so that works around clap stealing them :)

bbara replied:
Thanks for the hint, I guess the new location of the implementation fits better

tertsdiepraam commented on tests/by-util/test_tail.rs:
```diff
@@ -4479,3 +4479,232 @@ fn test_args_sleep_interval_when_illegal_argument_then_usage_error() {
     }
 }
+
+#[test]
```
Good idea to include the GNU tests! :)

tertsdiepraam commented:
> Congrats! The gnu test tests/misc/tail is no longer failing!

This is exciting!

bbara commented:
Not sure why `tests/tail-2/inotify-dir-recreate` fails. It passes on my machine and I couldn't find a log yet...

Also, I am running the cargo tests with `cargo test --features "tail" --no-default-features`, it seems like `parse::tests::test_parse_numbers_obsolete` (failed) is not included there.

tertsdiepraam commented:
> Not sure why tests/tail-2/inotify-dir-recreate fails. It passes on my machine and I couldn't find a log yet...

It might be just an intermittently failing test. If it keeps happening we can look into it, but I think it's unrelated to this change.

> Also, I am running the cargo tests with cargo test --features "tail" --no-default-features, it seems like parse::tests::test_parse_numbers_obsolete (failed) is not included there.

Yes, you need `--workspace` to also test the unit tests.

bbara commented:
Fine thx. I fixed the unit tests and now I'll wait for the code coverage report. I guess I have to add some parse unit tests after that.

tertsdiepraam commented on src/uu/tail/src/parse.rs:
```diff
@@ -78,15 +98,30 @@ pub fn parse_obsolete(src: &str) -> Option<Result<impl Iterator<Item = OsString>
                 }
                 Err(_) => Some(Err(ParseError::Overflow)),
             }
+        } else if src.len() == 2 {
```
I'm not sure these are corner cases per se. Looking at the syntax `-[num][bcl][f]`, I think the structure should be something like this:
- Parse the sign
- Parse optionally a number with suffix, defaulting to 10
- Parse optionally a mode, defaulting to `l`
- Parse optionally the `f` and set `-f` accoringly.

So there shouldn't be a special case for 2 characters, because that's just the case where the parsing of the number defaults to 10.

tertsdiepraam replied:
I just wrote basically the same function to test the new WIP argument parser. There are some differences (no sign parsing is required and it doesn't expand into other arguments), but it might still be helpful here: https://github.com/tertsdiepraam/uutils-args/blob/8616b87f5843ebafcb7312b204c673827fafbc09/tests/coreutils/tail.rs#L61-L94 (it hasn't been tested super extensively yet, so let me know if I made a mistake!)

bbara replied:
Looks good! I will take a deeper look into it tomorrow. There might be an ambiguity with e.g. `tail -c 20`. GNU's variant will handle it as `tail -20c`, but in our case it might result to "invalid file '20'".
And I have to check if `tail -[zqvf]` is handled by clap

tertsdiepraam replied:
I updated the parsing based on the things you mentioned: https://github.com/tertsdiepraam/uutils-args/blob/main/tests/coreutils/tail.rs

It's now very compatible I think. It even only accepts it as the first argument. I also don't need to deal with `-z`, `-q` and `-v`. Basically, I think that expanding to clap arguments was a mistake, because it makes the parsing more difficult and too lenient. It's much easier to just parse separately and fall back to clap if it fails.

tertsdiepraam replied:
(This doesn't need to happen in this PR though. What you've got looks like a good improvement!)

bbara replied:
As you have said, I guess clap will handle \-[fvqz]+ fine. I will remove the handling later today, s.t. it will be closer to your implementation.

bbara commented:
@tertsdiepraam should tail on x86 work with u64 input? because the implementation later (`Signum` struct) is working with u64. We could also switch to usize here, if not.

tertsdiepraam commented:
Well, it should not be up to the parsing to determine that. If that value is actually used as a `u64` then we should parse a `u64`, but if it's casted to a `usize` we should indeed parse a `usize`.

bbara commented:
Fine! The old implementation parsed the number as `usize`, therefore `test_parse_obsolete_overflow_x32` was also added. I removed it and parse it as `u64`, independent of the pointer-width.

tertsdiepraam commented:
Nice! Was `test_parse_obsolete_overflow_x32` based on a GNU test? If so, we could choose to ignore that test in the CI.

tertsdiepraam commented:
Wow, GNU is weird sometimes. What number does it use there? I think we can safely error on invalid utf-8. I can see people using invalid utf-8 for paths, but there's no reason to support it for numbers.

bbara commented:
Both weird:
```
$ tail -$'\x80'c
tail: invalid option -- ''
Try 'tail --help' for more information.
```
```
$ tail -c$'\x80'
tail: invalid number of bytes: ‘\200’
```

As you have mentioned it, the current (main) implementation does not allow non-UTF-8 paths:
https://github.com/uutils/coreutils/blob/d21448dc747f560c45f74ea56af041bae044de12/src/uu/tail/src/paths.rs#L30

I enabled it in my latest commit.

tertsdiepraam commented:
> ```
> $ tail -$'\x80'c
> tail: invalid option -- ''
> Try 'tail --help' for more information.
> ```

Here, it just fails to parse the deprecated syntax and falls back to normal parsing. We should do the same thing. The error message will be slightly different, because it will be a clap error, but that's fine.

> ```
> $ tail -c$'\x80'
> tail: invalid number of bytes: ‘\200’
> ```

This we can match if we want to. That's just displaying the same string but with octal, because 80 in hex is 200 in octal. However, using `to_string_lossy` is fine too in my opinion.

bbara commented:
The current state of the PR drops:
```
$ ./tail -$'\x80'c
./tail: bad argument encoding
```
```
$ ./tail -c$'\x80'
./tail: bad argument encoding
```

clap drops:
```
$ ./tail -v -c$'\x80'
error: Invalid UTF-8 was detected in one or more arguments

Usage: ./tail [FLAG]... [FILE]...

For more information try '--help'
```

similar run on GNU's tail:
```
$ tail -v -c$'\x80'
tail: invalid number of bytes: ‘\200’
```

I guess the current state of the PR is fine, what do you think?

tertsdiepraam commented:
The fact that it errors is fine, but it'd be great if the error was more descriptive. Maybe they could mention the option for which the argument was invalid and print some representation of the value (e.g. via `OsStr::display`).

bbara commented:
`OsStr::display` is not implemented, but I can use `to_string_lossy()` which would result e.g. to:
```
$ ./tail -$'\x80'b
./tail: bad argument encoding: -�b
```

tertsdiepraam commented:
> `OsStr::display` is not implemented, but I can use `to_string_lossy()` which would result e.g. to:
> 
> ```
> $ ./tail -$'\x80'b
> ./tail: bad argument encoding: '-�b'
> ```

Oh yeah that's fine! I was confused with `Path::display`, but that does the same thing.

bbara commented on src/uu/tail/src/args.rs:
```diff
@@ -357,9 +395,27 @@ fn parse_num(src: &str) -> Result<Signum, ParseSizeError> {
+pub fn parse_args(mut args: impl uucore::Args) -> UResult<Settings> {
+    let first = args.next().unwrap();
+    let second = args.next();
+    if let Some(second) = second {
```
Yep, much better. Yes, the MSRV doesn't enable it yet - failed some tests because of it :grin: I will adapt it. It's also easier to switch to the let else syntax, once the MSRV is high enough.

tertsdiepraam replied:
Yeah, I love `let ... else` and would bump the MSRV if rustfmt would support it. For now, it just crashes rustfmt which is a shame.

bbara commented on src/uu/tail/src/args.rs:
```diff
@@ -357,9 +395,27 @@ fn parse_num(src: &str) -> Result<Signum, ParseSizeError> {
+    let second = args.next();
+    if let Some(second) = second {
+        if let Some(second_str) = second.to_str() {
+            if let Some(obsolete_args) = parse_obsolete(second_str)? {
```
GNU drops specific error messages, therefore I guess we have to to be compliant. But what comes to my mind: I will try to first run clap and then the obsolete, as stated here:

`For compatibility tail also supports an obsolete usage ‘tail -[num][bcl][f] [file]’, which is recognized only if it does not conflict with the usage described above. `

tertsdiepraam replied:
Kinda, the only specific message is the overflow one, so I guess we indeed need this. For the `Context` one, that would actually be the error message from normal parsing I think.

bbara commented on src/uu/tail/src/parse.rs:
```diff
@@ -3,159 +3,134 @@
+    let mut follow = false;
+    let mut mode = 'l';
+    let mut first_char = true;
+    for char in chars.skip_while(|&c| c.is_ascii_digit()) {
```
I will adapt your implementation. Basically this is just a relict when also handling the `qvz` in this function.

bbara replied:
If you want, I can add a `Co-authored-by` to my last commit, as it is basically your implementation :)

tertsdiepraam replied:
Thanks, but there's no need. I think I got enough commits on this repo :)

bbara commented on src/uu/tail/src/args.rs:
```diff
@@ -128,6 +151,29 @@ pub struct Settings {
 }
 
 impl Settings {
+    pub fn from_obsolete_args(args: &parse::ObsoleteArgs, name: Option<OsString>) -> Self {
```
It does! :) `For compatibility tail also supports an obsolete usage ‘tail -[num][bcl][f] [file]’`

tertsdiepraam replied:
Oh yeah you're totally right!

bbara commented on src/uu/tail/src/parse.rs:
```diff
@@ -3,159 +3,134 @@
-                Err(_) => Some(Err(ParseError::Overflow)),
-            }
+/// tail -\[NUM\]\[bl\]\[f\] and tail +\[NUM\]\[bcl\]\[f\] // spell-checker:disable-line
+pub fn parse_obsolete(src: &str) -> Option<Result<ObsoleteArgs, ParseError>> {
```
I think the new one is even better :grinning:

bbara commented:
Is there a specific reason why the arguments are maintained as an iterator and not something better index- and rewind-able (e.g. something `Vec`-likely)? Not sure if needed in the other subprojects, but in `tail` this splitting and then re-chaining feels more complex than necessary.

tertsdiepraam commented:
It's probably because `std::env::args()` returns an iterator. If you want a vec you have to collect it first, which essentially means cloning it.

bbara commented:
I guess I am finished. What do you think @tertsdiepraam? :)

tertsdiepraam commented on src/uu/tail/src/args.rs:
```diff
@@ -358,8 +421,35 @@ fn parse_num(src: &str) -> Result<Signum, ParseSizeError> {
+    let clap_result = match uu_app().try_get_matches_from(args_vec.clone()) {
+        Ok(matches) => {
+            let settings = Settings::from(&matches)?;
+            if !settings.is_default() {
+                // non-default settings can't have obsolete arguments
+                return Ok(settings);
+            }
```
Why can default settings have obsolete arguments? Seems to me that if clap succeeds we should always use that value? What is a case where that's not true?

bbara replied:
> ```shell
> tail -f -f
> ```
> 
> Should try to read the file called `-f`. 

> 
> ```shell
> tail -f -f -f
> ```
> 
> should be allowed. 

No, in GNU's implementation both activate -f and use STDIN as input.

> If we fix that by adding `.overrides_with(option::FOLLOW)` to clap. 

IMHO, we should allow overrides, as this is what GNU does. `tail (-f)+` is basically the same `tail -f --`, no matter how much `-f` follow. 

> Then `-f -f` is not parsed as obsolete syntax in your implementation.

Obsolete should only be parsed for following format: `tail <args> <file>?` (where args can be `-f+`).
so `tail -f -f -f` would basically be `tail -fff` in obsolete syntax, semantically resulting to `tail -f --`.

bbara replied:
AH, wrong window...

Answer to this: Problem is e.g. `tail +c`. This is handled as valid by clap (default tail to file `+c`).

tertsdiepraam replied:
Ohhh you're totally right! Sorry!

tertsdiepraam commented on src/uu/tail/src/args.rs:
```diff
@@ -302,34 +354,45 @@ impl Settings {
+                        arg_str.quote()
+                    ),
+                    parse::ParseError::Overflow => format!("invalid number: {}", arg_str.quote()),
+                    parse::ParseError::Context => format!(
+                        "option used in invalid context -- {}",
+                        arg_str.chars().nth(1).unwrap_or_default()
+                    ),
```
Wouldn't this be caught by clap already?

bbara replied:
It would! But clap's error message is different:

```
$ ./target/release/tail -5cz
error: Found argument '-5' which wasn't expected, or isn't valid in this context

  If you tried to supply '-5' as a value rather than a flag, use '-- -5'

Usage: ./target/release/tail [FLAG]... [FILE]...

For more information try '--help'
```

vs

```
$ tail -5cz
tail: option used in invalid context -- 5
```

This is relevant to parse the misc/tail test of GNU (checks for the error output).

tertsdiepraam replied:
Yes, but we tend to not be compatible with error messages in general. Instead we can patch the GNU test suite. This is part of what https://github.com/uutils/coreutils/blob/main/util/build-gnu.sh does. I'd say that this PR is ready if it passes all the tests you copied with error messages adapted to clap where necessary.

bbara replied:
I can remove it if the clap error is generally preferred. But I think in this case we kind of get it "for free". But obviously, when returning None instead of the Error, we will see clap's error instead.

tertsdiepraam replied:
Alright, we'll leave it in, but maybe you could add a comment explaining why it's there?

tertsdiepraam commented on src/uu/tail/src/args.rs:
```diff
@@ -358,8 +421,35 @@ fn parse_num(src: &str) -> Result<Signum, ParseSizeError> {
+        Some(second) => second,
+        None => return clap_result,
+    };
+    match parse_obsolete(second)? {
```
Maybe it's easier if `parse_obsolete` just returns `UResult<Option<Settings>>`?

tertsdiepraam commented on src/uu/tail/src/args.rs:
```diff
@@ -302,34 +354,45 @@ impl Settings {
+            && self.pid == default.pid
+            && self.retry == default.retry
+            && self.use_polling == default.use_polling
+            && (self.verbose == default.verbose || self.inputs.len() > 1)
```
Maybe I'm being dumb but this doesn't make sense to me. What's your reasoning here? It is the default if there are multiple inputs? Shouldn't the inputs be the same as the default as well? Ideally this would just be a `#[derive(Eq)]`, of course, but maybe that's not possible?

bbara replied:
Yes. Verbose is automatically added, when there is more than one input:
https://github.com/uutils/coreutils/blob/4e7ae2d46a8550e7bd372a80449662fd646e895e/src/uu/tail/src/args.rs#L221

Basically, the reason is that in calls with obsolete arguments, like `tail +2c`, clap interprets it as "default tail on file `+2c`". My workaround from above, where this is needed, should do something like "if you use the default arguments, please check the first input if it isn't a obsolete argument (instead of a non-existing file (which is not known at this point))".

Therefore Eq is not possible, it would also include the inputs, which are not part of the "settings" (as interpretation of the args).

But I guess I have to add some comments to clarify? :grin:

tertsdiepraam replied:
I don't love it to be honest. It seems kind of like a brittle way of doing things.

If it's for the `+` specifically, what about maybe this logic?

1. Try clap
2. If it fails, try obsolete
3. If it succeeds, check the original arguments for whether the length is two and the first character is `+` and try obsolete if that's both true.

bbara replied:
Yup I find it ugly too. I will look for the cases not covered by clap (e.g. single input with starting +) and run the obsolete parsing on it tomorrow

bbara replied:
> If it's for the `+` specifically, what about maybe this logic?
>     3. If it succeeds, check the original arguments for whether the length is two and the first character is `+` and try obsolete if that's both true.

Unfortunately not. At least `tail -<num>` is also handled by the obsolete parsing.

bbara replied:
IMHO, the main problem is to decide ambiguous calls, like e.g. `tail -c +3`:
Parsed by clap, it means "run tail with c=+3 on STDIN".
Parsed by obsolete, it means "run tail with c=-10 on +3".

My latest version checks if clap found only inputs, and if so, runs the obsolete handling afterwards. What do you think?

tertsdiepraam replied:
> Unfortunately not. At least tail -<num> is also handled by the obsolete parsing.

Doesn't step 2 take case of that? `-<num>` should fail in clap right?

bbara replied:
Obviously your solution would also work, and I forgot to implement step 2 yesterday while testing... It's basically the same condition but I think yours is simpler. I'll check later

tertsdiepraam replied:
> Obviously your solution would also work

Wasn't so obvious to me :) I'm glad you're doing so much extensive testing

tertsdiepraam reviewed (commented):
I found another cursed edge case:
```sh
tail -f -f
```
Should try to read the file called `-f`. Now technically, your implementation does this correctly. However, that is due to another bug, because the current implementation doesn't allow `-f` to override itself. For example,
```sh
tail -f -f -f
```
should be allowed. If we fix that by adding `.overrides_with(option::FOLLOW)` to clap. Then `-f -f` is not parsed as obsolete syntax in your implementation.

This shows that the obsolete syntax has priority over parsing in clap, despite what the documentation states. So I think we should return to trying to parse the obsolete syntax before clap.

bbara commented on src/uu/tail/src/args.rs:
```diff
@@ -358,8 +408,32 @@ fn parse_num(src: &str) -> Result<Signum, ParseSizeError> {
+        Err(err) => Err(err.into()),
+    };
+
+    // clap misinteprets obsolete args as first input (starting with +) or Err
```
oh no, typo :(

tertsdiepraam commented on src/uu/tail/src/args.rs:
```diff
@@ -358,8 +408,32 @@ fn parse_num(src: &str) -> Result<Signum, ParseSizeError> {
+    // clap misinterprets obsolete args as first input (starting with +) or Err
+    if let Ok(settings) = clap_result.as_ref() {
+        // Settings::from() ensures we always have at least one input (STDIN as fallback)
+        let input = settings.inputs.get(0).unwrap();
```
Instead of `.get(0).unwrap()`, you can just use `settings.inputs[0]`.

tertsdiepraam replied:
Also, I think the best value to check here is `args_vec.get(1)`, because the first input is not necessarily guaranteed to be the first argument. (Though I can't think of a case where your current implementation fails, it's just a bit safer)

tertsdiepraam commented on src/uu/tail/src/parse.rs:
```diff
@@ -5,157 +5,185 @@
-                }
-                Err(_) => Some(Err(ParseError::Overflow)),
-            }
+/// tail -\[NUM\]\[bl\]\[f\] and tail +\[NUM\]\[bcl\]\[f\] // spell-checker:disable-line
```
In this comment, `c` should also be allowed for `-` right?

bbara replied:
yes, true. I just didn't add it, because "tail -c" is handled by clap (valid flag), but "tail -10c" isn't. But I will add it, makes sense

tertsdiepraam commented on src/uu/tail/src/args.rs:
```diff
@@ -358,8 +408,32 @@ fn parse_num(src: &str) -> Result<Signum, ParseSizeError> {
+        Err(err) => Err(err.into()),
+    };
+
+    // clap misinterprets obsolete args as first input (starting with +) or Err
```
I think the logic is good, but let's expand the explanation a bit. Something along these lines (feel free to adapt or expand further):

> There are a few cases of ambiguity between clap parsing and the obsolete syntax:
>  1. Cases like `tail -f foo` and `tail -c 5`, where the first argument is both a valid flag and a valid first argument of the obsolete syntax. These should be handled as normal clap arguments.
> 2.  Cases like `tail +f bar` where `+f` is parsed as a filename by clap. In this case, clap succeeds, but we still want to discard the result in favor of obsolete parsing. So below, we check whether the first argument starts with `+` even when clap succeeds.

tertsdiepraam commented on src/uu/tail/src/paths.rs:
```diff
@@ -26,21 +27,21 @@ pub struct Input {
-        let kind = if string == text::DASH {
+    pub fn from<T: AsRef<OsStr>>(string: &T) -> Self {
+        let valid_string = string.as_ref().to_str();
+        let kind = if valid_string.is_some() && valid_string.unwrap() == text::DASH {
```
An easier check here is `string == Path::new(text::DASH)`

tertsdiepraam reviewed (commented):
Nice! We're getting very close to done here :)

tertsdiepraam commented on src/uu/tail/src/args.rs:
```diff
@@ -358,8 +409,37 @@ fn parse_num(src: &str) -> Result<Signum, ParseSizeError> {
+        return clap_result;
+    }
+
+    // there are cases of ambiguity between clap's parsing and obsolete syntax parsing.
+    // e.g. "tail -c 5", which is a valid flag (handled by clap),
+    // but could also be interpreted as valid obsolete syntax (tail -c on file '5').
+    // in this case, the clap result should be returned immediately.
+    // clap fails for unknown flags, but valid obsolete arguments (e.g. tail -l; tail -10c).
+    // clap succeeds for obsolete arguments starting with '+', but misinterprets them as
+    // input files (e.g. tail +f).
+    // in these cases, the argv[1] should be checked further.
```
Nice, I like that you explained more cases than I did, but let's bring back some of the formatting I had in my version and fix some of the punctuation. I tried to combine our approaches (again, feel free to improve further):

```rust
// At this point, there are a few possible cases:
//
//    1. clap has succeeded and the arguments would be invalid for the obsolete syntax.
//    2. The case of `tail -c 5` is ambiguous. clap parses this as `tail -c5`,
//       but it could also be interpreted as valid obsolete syntax (tail -c on file '5').
//       GNU chooses to interpret this as `tail -c5`, like clap.
//    3. `tail -f foo` is also ambiguous, but has the same effect in both cases. We can safely
//        use the clap result here.
//    4. clap succeeded for obsolete arguments starting with '+', but misinterprets them as
//       input files (e.g. tail +f).
//    5. clap failed because of unknown flags, but possibly valid obsolete arguments
//        (e.g. tail -l; tail -10c).
//
// In cases 4 & 5, we want to try parsing the obsolete arguments, which corresponds to
// checking whether clap succeeded or the first argument starts with '+'.
```

bbara replied:
Thanks for helping out with the format!

tertsdiepraam reviewed (approved):
Awesome! That's another GNU test passing. Thank you!
